src/scanpy/metrics/_metrics.py

```diff
@@ -89,3 +95,61 @@ def confusion_matrix(
     df = df.loc[np.array(orig_idx), np.array(new_idx)]
 
     return df
+
+
+def modularity(connectivities, labels, mode="UNDIRECTED") -> float:
+    # default mode is undirected?? can be specified as directed or undirected
+    """Compute the modularity of a graph given its connectivities and labels.
+
+    Parameters
+    ----------
+    connectivities: array-like or sparse matrix
+        Weighted adjacency matrix representing the graph. Can be a dense NumPy array or a sparse CSR matrix.
+    labels: array-like or pandas.Series
+        Cluster labels for each node in the graph.
+    mode: str
+        The mode of the graph. Can be "UNDIRECTED" or "DIRECTED". Default is "UNDIRECTED".
+
+    Returns
+    -------
+    float
+        The modularity of the graph based on the provided clustering.
+    """
+    if isinstance(connectivities, CSRBase):
+        # Convert sparse matrix to dense format so that igraph can handle it
+        # Weighted_Adjacency expects with nested lists or numpy arrays and not sparse matrices
+        dense_connectivities = connectivities.toarray()
+    else:
+        dense_connectivities = connectivities
+    # creating igraph graph from the dense connectivity matrix
+    graph = ig.Graph.Weighted_Adjacency(dense_connectivities.tolist(), mode=mode)
+    if isinstance(labels, pd.Series):
+        labels = labels.values
+    # making sure labels are in the right format, i.e., a list of integers
+    labels = pd.Categorical(np.asarray(labels)).codes
```

Comment:
that’s a lot of converting around! Are you sure you can’t just pass `labels` into `pd.Categorical` directly?